### PR TITLE
Recipe ai with preference

### PR DESCRIPTION
**Pull request by mvictorica97**
#### Board:
* [Ticket #NUMBER_OF_THE_TICKET](link_goes_here)
---
#### Description:

---
#### Notes:
*
---
#### Tasks:
- [x] Add each element in this format
---
#### Risk:
*
---

### CODE REVIEW

File: app/services/recipe_generator_service.rb [hunk 1]
```diff
@@ -41,9 +41,33 @@ def system_message
   end
 
   def prompt
-    <<~CONTENT
-      Prompt goes here
-    CONTENT
+    preferences_text = format_preferences
```
Comment (lauperalti) [inline app/services/recipe_generator_service.rb]: I think this variable `preferences_text` definition is not necessary, you could just use `format_preferences` below.

File: app/services/recipe_generator_service.rb [hunk 2]
```diff
@@ -41,9 +41,33 @@ def system_message
...
+  If any ingredient conflicts with these restrictions, exclude it from the recipe.
+  
+  You are expected to send a JSON in the response with the following format: 
+  {
+    "name": "Dish Name",
+    "content": ""
+  }
+
+  Ensure the recipe respects all restrictions and preferences. Please respond with the recipe details in Spanish.
+CONTENT
```
Comment (lauperalti) [inline app/services/recipe_generator_service.rb]: This is great!!!

File: app/services/recipe_generator_service.rb [hunk 3]
```diff
@@ -41,9 +41,33 @@ def system_message
...
+
+    Rails.logger.info "Generated prompt: #{prompt_content}"
```
Comment (lauperalti) [inline app/services/recipe_generator_service.rb]: I guess this was for testing purposes, not needed.

File: app/services/recipe_generator_service.rb [hunk 4]
```diff
@@ -41,9 +41,33 @@ def system_message
...
+    prompt_content
+  end
+
+  #formateo las preferences en una cadena para enviar al prompt
```
Comment (lauperalti) [inline app/services/recipe_generator_service.rb]: In rails we do not use much comments, code should be self-explanatory

File: app/services/recipe_generator_service.rb [hunk 5]
```diff
@@ -57,10 +81,43 @@ def openai_client
...
+      ingredients: message,
+      user_id: user.id
+    )
+  
+    if recipe.save
+      return recipe
+    else
+      raise RecipeGeneratorServiceError, 'Failed to create recipe.'
+    end
+  end
```
Comment (lauperalti) [inline app/services/recipe_generator_service.rb]: Long methods are not good practice in rails. if you have a long function you may need to think of splitting it.

Check [this Sandi Metz rules](https://thoughtbot.com/blog/sandi-metz-rules-for-developers): "Methods can be no longer than five lines of code." 😅

**Review by lauperalti — COMMENTED** [review]
This is great progress! 👏 👏 

I like how you build a very detailed prompt ✅  
Also the Service call from the controller is great ✅ 
The way you handle errors by raise and rescue RecipeGeneratorServiceError is really nice ✅ 

The method `create_recipe` has good validations and conditions, but it would be better to split it and have a more clean and clear solution. 🙏

File: app/services/recipe_generator_service.rb (hunk 1 above)
Comment (mvictorica97) [reply app/services/recipe_generator_service.rb]: Deleted the variable definition and called the function format_preferences directly in the prompt
`User preferences: #{format_preferences}.`

Also deleted `    Rails.logger.info "Generated prompt: #{prompt_content}"
    prompt_content` in lines 58 and 59 which where used to debug the prompt.

File: app/services/recipe_generator_service.rb (hunk 4 above)
Comment (mvictorica97) [reply app/services/recipe_generator_service.rb]: Deleted all comments, and changed variable/function names so its easier to understand what everything does.

File: app/services/recipe_generator_service.rb (hunk 5 above)
Comment (mvictorica97) [reply app/services/recipe_generator_service.rb]: To make create_recipe a shorter method, I am handling the recipe creation in a new method:
`def recipe_info(name, content, message)
    Recipe.new(
      name: name,
      description: content,
      ingredients: message,
      user_id: user.id
    )
 end`

I am calling the recipe_info method in the create_recipe method this way:
`recipe = recipe_info(name, content, message)`

**Pull request merged**